### ISSUE TITLE
`clean` task always fails

### ISSUE THREAD

**Issue by henkelmax** [neoforged/NeoGradle#18]
Running the `clean` gradle task always results in this error:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':clean'.
> java.io.IOException: Unable to delete directory 'E:\Workspaces\MDK\build'
    Failed to delete some children. This might happen because a process has files open or has its working directory set in the target directory.
    - E:\Workspaces\MDK\build\tmp\.cache\expanded\expanded.lock
    - E:\Workspaces\MDK\build\tmp\.cache\expanded
    - E:\Workspaces\MDK\build\tmp\.cache
    - E:\Workspaces\MDK\build\tmp
```

This even happens with the [MDK](https://github.com/neoforged/MDK) project.

**Comment by mysticdrew** [comment]
I would run clean, then delete that lock file manually. Not a good solution though.

**Comment by henkelmax** [comment]
> I would run clean, then delete that lock file manually. Not a good solution though.

Yeah I just manually deleted the build folder instead of running the clean task for now (I think thats just what the task does).

**Comment by marchermans** [comment]
I currently can't reproduce this.

**Comment by henkelmax** [comment]
> I currently can't reproduce this.

1. Clone the [MDK](https://github.com/neoforged/MDK) repo
2. Open a command line
3. Enter `./gradlew clean`
4. See error

Also happens when running the clean task from intellij

**Comment by marchermans** [comment]
> > I currently can't reproduce this.
> 
> 1. Clone the [MDK](https://github.com/neoforged/MDK) repo
> 2. Open a command line
> 3. Enter `./gradlew clean`
> 4. See error
> 
> Also happens when running the clean task from intellij

That is literally what I did

**Comment by henkelmax** [comment]
> That is literally what I did

Really weird...
Maybe its operating system or file system specific. For me it happens every time I do it.

**Comment by marchermans** [comment]
![image](https://github.com/neoforged/NeoGradle/assets/5585406/73214d77-959b-4cf1-bea4-ce31168eb5fc)

**Comment by marchermans** [comment]
Can you try to figure out what process is holding that lock?

**Comment by henkelmax** [comment]
> Can you try to figure out what process is holding that lock?

It is the Gradle process itself. Even if the lock file doesn't exists, it creates it while cleaning.

![image](https://github.com/neoforged/NeoGradle/assets/13237524/bf0c4dbf-d5d0-400b-b05f-2a03469be74a)

Maybe its a race condition that only happens on some systems.

**Comment by sciwhiz12** [comment]
I've just had the same bug hit me while running `clean` on my NeoForge workspace -- both `:base` and `:neoforge` got hit by the bug, and for the same file/directory tree of `build\tmp\.cache\expanded\expanded.lock` (relative to the project directory).

**Comment by marchermans** [comment]
> I've just had the same bug hit me while running `clean` on my NeoForge workspace -- both `:base` and `:neoforge` got hit by the bug, and for the same file/directory tree of `build\tmp\.cache\expanded\expanded.lock` (relative to the project directory).

Interesting could this be a Gradle bug?

@sciwhiz12 or @henkelmax try to reproduce it without the daemon, maybe?

**Comment by henkelmax** [comment]
> Interesting could this be a Gradle bug?
>
> @sciwhiz12 or @henkelmax try to reproduce it without the daemon, maybe?

```
./gradlew clean --no-daemon
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/8.1.1/userguide/gradle_daemon.html#sec:disabling_the_daemon.
Daemon will be stopped at the end of the build
> Task :clean FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':clean'.
> java.io.IOException: Unable to delete directory 'E:\Workspaces\MDK\build'
    Failed to delete some children. This might happen because a process has files open or has its working directory set in the target directory.
    - E:\Workspaces\MDK\build\tmp\.cache\expanded\expanded.lock
    - E:\Workspaces\MDK\build\tmp\.cache\expanded
    - E:\Workspaces\MDK\build\tmp\.cache
    - E:\Workspaces\MDK\build\tmp

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 11s
1 actionable task: 1 executed
```

Same issue unfortunately

**Comment by Technici4n** [comment]
Does this happen if you first kill all deamons with `gradlew --stop` and then try `gradlew clean`?

**Comment by henkelmax** [comment]
> Does this happen if you first kill all deamons with `gradlew --stop` and then try `gradlew clean`?

Same result

**Comment by Technici4n** [comment]
It is probably an upstream Gradle issue. See https://github.com/gradle/gradle/issues/25752#issuecomment-1792821355.

This only happens on Windows.

**Comment by Leialoha** [comment]
For anyone still experiencing this error, here is a workaround:

```groovy
// Removing all files but the expanded.lock
task customClean(type: Delete) {
    // Specify the files or directories to delete
    // delete 'otherDirectory'

    // Remove all files except that one lock file
    def files = []

    fileTree("build")
    .exclude('**/expanded.lock')
    .visit { FileVisitDetails fvd ->
        files << fvd.file
    }
    
    files.reverse()
    .each { f ->
        def file = file(f)
        if (!file.exists()) return;
        if ((file.isDirectory() && file.list().size() == 0)
            || !file.isDirectory())
                delete f
    }

}

clean.dependsOn customClean
clean.enabled = false
```

**Comment by henkelmax** [comment]
> For anyone still experiencing this error, here is a workaround:

This script threw an error for me.
Wrapping your script in 
``` groovy
doLast {
  // ...
}
```
worked though.

<details>
  <summary>Full script</summary>

  ```groovy
  tasks.register('customClean', Delete) {
    doLast {
        def files = []
        fileTree("build")
                .exclude('**/expanded.lock')
                .visit { FileVisitDetails fvd ->
                    files << fvd.file
                }
        files.reverse()
                .each { f ->
                    def file = file(f)
                    if (!file.exists()) return;
                    if ((file.isDirectory() && file.list().size() == 0)
                            || !file.isDirectory())
                        delete f
                }
    }
}
clean.dependsOn customClean
clean.enabled = false
  ```
  
</details>

**Comment by marchermans** [comment]
> For anyone still experiencing this error, here is a workaround:
> 
> ```groovy
> // Removing all files but the expanded.lock
> task customClean(type: Delete) {
>     // Specify the files or directories to delete
>     // delete 'otherDirectory'
> 
>     // Remove all files except that one lock file
>     def files = []
> 
>     fileTree("build")
>     .exclude('**/expanded.lock')
>     .visit { FileVisitDetails fvd ->
>         files << fvd.file
>     }
>     
>     files.reverse()
>     .each { f ->
>         def file = file(f)
>         if (!file.exists()) return;
>         if ((file.isDirectory() && file.list().size() == 0)
>             || !file.isDirectory())
>                 delete f
>     }
> 
> }
> 
> clean.dependsOn customClean
> clean.enabled = false
> ```

Interesting.... @Technici4n Maybe report this to gradle as well.

**Comment by Technici4n** [comment]
They are already working on a proper fix for 8.6 ;)

https://github.com/gradle/gradle/pull/27392

**Comment by UpcraftLP** [comment]
> They are already working on a proper fix for 8.6 ;)
> 
> [gradle/gradle#27392](https://github.com/gradle/gradle/pull/27392)

Update: [as stated on the gradle issue discussion](https://github.com/gradle/gradle/issues/25752#issuecomment-1884784452), this will be fixed in gradle 8.7.

**Comment by KlemenDEV** [comment]
Gradle 8.7 is out and clean seems to work now :)
